### ISSUE TITLE
Quality of even 1 iteration of model.fit is strongly depends on nb_epoch

### ISSUE THREAD

**Issue by MaratZakirov** [fchollet/keras#5538]
**Please pay attention to this because user experience here is completely unusual.**

I am asking because model.fit(nb_epoch=1) and model.fit(nb_epoch=10) is my case gives dramatically different results (after just one iteration) when model.fit(nb_epoch=10) is much better. I also interested in model.fit_generator, it also behaves different from model.fit (fit_generator is much worse and it is also creates FPE very often). So the problem is seems to be in learning rate adjustments. Do anybody have explanation to that?

```
model.fit(x=Q, y=L, validation_split=0.1, nb_epoch=1)
Train on 450000 samples, validate on 50000 samples
Epoch 1/1
450000/450000 [==============================] - 40s - loss: 0.0432 - val_loss: 0.0389

model.fit(x=Q, y=L, validation_split=0.1)#, nb_epoch=1)
Train on 450000 samples, validate on 50000 samples
Epoch 1/10
450000/450000 [==============================] - 40s - loss: 0.0219 - val_loss: 0.0141
Epoch 2/10
119584/450000 [======>.......................] - ETA: 28s - loss: 0.0148
```
I also consider for now fit_generator as **completely unusable thing**.

**Comment by kgrm** [comment]
`fit_generator` obviously depends strongly on the generator you're feeding it with. You should also note that the default behaviour of `model.fit` is random batch shuffling, in which case the performance after a single epoch will mostly be a function of the batch order, and you can only reasonably compare performance after several such epochs. If that's not what you want, use a deterministic batch order to compare. And if you want fully reproducible behaviour, you'll also need to fix model weight initialisation and the random generator seeds.

**Comment by unrealwill** [comment]
Hello,

Are you doing 2 calls one after the other to model.fit?

Because as @kgrm said, the weights initialization must be the same. Here the first model.fit modify the weights. So the second model.fit doesn't start from scratch (And indeed obtain a better loss).

You can use save_weights and load_weights to save and restore weights and shuffle=False so that you obtain a fully reproducible behaviour.

**Comment by MaratZakirov** [comment]
@unrealwill 

No!!!! Every experiment is from the scratch

```
#  -*- coding: utf-8 -*-
from keras.models import Sequential
from keras.layers import Dropout, Dense, RepeatVector, LSTM, Masking, TimeDistributed, Embedding, Bidirectional, SimpleRNN

# from keras.utils.visualize_util import plot
from keras.layers import Merge, Activation
import numpy
import theano.tensor as T
from sklearn import decomposition

def linesnum(file):
    cnt = 0
    for line in open(file):
        cnt += 1
    return cnt

def loadtable(file):
    table = {}
    first = True
    for line in open(file):
        if first:
            first = False
            continue
        words = line.rstrip().split()
        table[words[0]] = numpy.array(words[1:]).astype('float32')
    return table

def loadGen(file, table, q_seq_size, is_train=True, batch_size=-1):#500000):
    featnum = table.values()[0].shape[0]
    Q = numpy.zeros(shape=(batch_size, q_seq_size, featnum), dtype='float32')
    L = numpy.zeros(shape=batch_size, dtype='float32')
    batch_ind = 0
    while True:
        for line in open(file):
            parts = line.rstrip().split('\t')
            j = 0
            if is_train:
                L[batch_ind] = float(parts[0])
                words = parts[1].split()
            else:
                words = parts[0].split()
            for word in words:
                if word not in table: continue
                if j >= q_seq_size: break
                Q[batch_ind][j] = table[word]
                j += 1
            batch_ind += 1
            if batch_ind == batch_size:
                batch_ind = 0
                if is_train: yield (Q, L)
                else:        yield Q
                Q = numpy.zeros(shape=(batch_size, q_seq_size, featnum), dtype='float32')
        if not is_train: yield Q[0:batch_ind]
        else:            yield (Q[0:batch_ind], L[0:batch_ind])
            
def loadData(nfile, table, q_seq_size):
    featnum = table.values()[0].shape[0]
    batch_size = linesnum(nfile)
    Q = numpy.zeros(shape=(batch_size, q_seq_size, featnum), dtype='float32')
    L = numpy.zeros(shape=batch_size, dtype='float32')
    line_id = 0
    for line in open(nfile):
        parts = line.rstrip().split('\t')
        L[line_id] = float(parts[0])
        j = 0
        words = parts[1].split()
        for word in words:
            if word not in table: continue
            if j >= q_seq_size: break
            Q[line_id][j] = table[word]
            j += 1
        line_id += 1  
    return Q, L

def train(filetrain, filevalid, table, q_seq_size=10, TEST = '/home/zakirov/data/epgu-class/test1.txt'):
    featnum = table.values()[0].shape[0]
    # define positive rnn convolution
    model = Sequential()
    model.add(SimpleRNN(20, input_dim=featnum))
    model.add(Dense(1, activation='sigmoid'))
    model.compile(optimizer='adam', loss='mse')
    print 'Number of model parameters: ', model.count_params()
    if 0:
        Q, L = loadData(filetrain, table, q_seq_size)
        model.fit(x=Q, y=L, validation_split=0.1)#, nb_epoch=1)
    elif 0:
        train_size = linesnum(filetrain)
        valid_size = linesnum(filevalid)
        model.fit_generator(generator=loadGen(filetrain, table, q_seq_size, batch_size=train_size),
                            validation_data=loadGen(filevalid, table, q_seq_size, batch_size=valid_size),
                            nb_epoch=100,
                            samples_per_epoch=train_size,
                            nb_val_samples=valid_size)
    else:
        for Q, L in loadGen(filetrain, table, q_seq_size, batch_size=500000):
            model.fit(x=Q, y=L, validation_split=0.1)

    Qt, Lt = loadData(TEST, table, q_seq_size)
    model.predict(Qt, Lt)        
    return model

VALID = '/home/zakirov/data/epgu-class/valid3.txt'
TRAIN = '/home/zakirov/data/epgu-class/learn3.txt'
TABLE = '/home/zakirov/data/epgu-class/vec-epgu.100'

table = loadtable(TABLE)
print 'Table loaded'
model, conf = train(TRAIN, VALID, table)
```

**Comment by unrealwill** [comment]
```
 else:
        for Q, L in loadGen(filetrain, table, q_seq_size, batch_size=500000):
            model.fit(x=Q, y=L, validation_split=0.1)
```

You are mixing "generator" and "fit".  By default fit has 10 epochs. 
So you are trying to manually do a fit_generator, but for each Q,L you are looping 10 times. Before moving to the next Q, L. (By the way, the "epoch" displayed doesn't correspond to a "full epoch" (i.e. iterating over the whole data), because you are doing things manually and by passing it a smaller dataset model.fit has no way of knowing it.).


Also your loadGen fonction is buggy. (There may be more bugs I didn't tried to run your code)

```
        if not is_train: yield Q[0:batch_ind]
        else:            yield (Q[0:batch_ind], L[0:batch_ind])
```
Is not properly indented and should be inside the " if batch_ind == batch_size: " branch.

**Comment by MaratZakirov** [comment]
@unrealwill I want to thank you for code review (of the code I just deleted due to its complexity).
I want you to think about more precise problem:
model.fit(x=Q, y=L, validation_split=0.1, nb_epoch=1) versus model.fit(x=Q, y=L, validation_split=0.1, nb_epoch=10). Their behaviour is different

**Comment by unrealwill** [comment]
Of course they are different. In one case you are doing x updates, and in the other case you are doing 10 x updates.

**Comment by MaratZakirov** [comment]
@unrealwill 
There is a misleading. I do not understand my concern.
Take some code:
```
def train(filetrain, filevalid, table, q_seq_size=10, TEST = '/home/zakirov/data/epgu-class/test1.txt'):
    featnum = table.values()[0].shape[0]
    # define positive rnn convolution
    model = Sequential()
    model.add(SimpleRNN(20, input_dim=featnum))
    model.add(Dense(1, activation='sigmoid'))
    model.compile(optimizer='adam', loss='mse')
    print 'Number of model parameters: ', model.count_params()
    Q, L = loadData(filetrain, table, q_seq_size)
    model.fit(x=Q, y=L, validation_split=0.1, nb_epoch=1)
```
If I setup "nb_epoch=1" I get val_loss: 0.0389
If I setup "nb_epoch=10" I get val_loss: 0.0141

All at the SAME data

**Comment by unrealwill** [comment]
What you were doing was (over)fitting to a sub-dataset, by looping 10 times on it. Then you moved to the next sub-dataset, and looped 10 times on it (and therefore forget the first dataset)

**Comment by unrealwill** [comment]
You still get a better val_loss in the second case because you did 10 times more updates during your "one" pass of the data.

**Comment by MaratZakirov** [comment]
```
        Q, L = loadData(filetrain, table, q_seq_size)
        model.fit(x=Q, y=L, validation_split=0.1, nb_epoch=5)
Train on 450000 samples, validate on 50000 samples
Epoch 1/5
450000/450000 [==============================] - 42s - loss: 0.0283 - val_loss: 0.0195
Epoch 2/5
450000/450000 [==============================] - 42s - loss: 0.0169 - val_loss: 0.0144
Epoch 3/5
450000/450000 [==============================] - 42s - loss: 0.0142 - val_loss: 0.0131
Epoch 4/5
450000/450000 [==============================] - 42s - loss: 0.0128 - val_loss: 0.0132
Epoch 5/5
450000/450000 [==============================] - 42s - loss: 0.0121 - val_loss: 0.0120
```

```
        Q, L = loadData(filetrain, table, q_seq_size)
        model.fit(x=Q, y=L, validation_split=0.1, nb_epoch=1)
        model.fit(x=Q, y=L, validation_split=0.1, nb_epoch=1)
        model.fit(x=Q, y=L, validation_split=0.1, nb_epoch=1)
        model.fit(x=Q, y=L, validation_split=0.1, nb_epoch=1)
        model.fit(x=Q, y=L, validation_split=0.1, nb_epoch=1)

Train on 450000 samples, validate on 50000 samples
Epoch 1/1
450000/450000 [==============================] - 42s - loss: 0.0386 - val_loss: 0.0353
Train on 450000 samples, validate on 50000 samples
Epoch 1/1
450000/450000 [==============================] - 41s - loss: 0.0362 - val_loss: 0.0443
Train on 450000 samples, validate on 50000 samples
Epoch 1/1
450000/450000 [==============================] - 41s - loss: 0.0415 - val_loss: 0.0296
Train on 450000 samples, validate on 50000 samples
Epoch 1/1
450000/450000 [==============================] - 40s - loss: 0.0289 - val_loss: 0.0269
Train on 450000 samples, validate on 50000 samples
Epoch 1/1
450000/450000 [==============================] - 40s - loss: 0.0277 - val_loss: 0.0283
```

**Comment by unrealwill** [comment]
Try doing it with shuffle = False, inside fit, and this should get the same results.

**Comment by MaratZakirov** [comment]
It doesn't help. Seems to me that it is some kind of random effect with probability depending on some specific model.fit options. But it actually OK just not best performance is certain cases. The real (WTF kind of) problem is fit_generator

**Comment by unrealwill** [comment]
There is no problem in either fit or fit_generator (it's kind of the core of the library so it would be have been noticed by now).

The probable reason for which fit_generator may sometimes work poorly (other than obvious bugs) is when you don't shuffle the data (which you must write inside your generator (which I don't think you had) ) (It's particularly problematic when there is a strong correlation between adjacent example).

**Comment by MaratZakirov** [comment]
@unrealwill The reality is that I gave fit_generator batch size of WHOLE DATA and it is still have WTF

**Comment by unrealwill** [comment]
Then that's the problem. You shouldn't give the WHOLE DATA to fit generator. 
Otherwise it does only one update by epoch.
You should give it a mini_batch. 
Fit automatically uses batch_size=32.
When you implement your generator you should yield a mini batch.

**Comment by MaratZakirov** [comment]
@unrealwill  If I set small size of data to pick up at a time it works pretty slow and nb_worker is a deprecated option. I also do not think that 512 of batch size (which I usually use in my generator) is not  bad enough to devastate learning process.

**Comment by MaratZakirov** [comment]
It is seems that fit_generator works improperly when using loss='mse' I will file another issue latter
